#+STARTUP: overview
#+TITLE: HSPC dynamics analysis code repository

This repository contains R, python and Matlab scripts used to perform analysis published in the Kucinski et al. 2023 paper: "A time and single-cell resolved model of murine bone marrow hematopoiesis".

An interactive visualisation can be found: [[https://gottgens-lab.stemcells.cam.ac.uk/bgweb2/HSPC_dyn2022/][here]]

* scRNA-Seq analysis
All the scripts for analysis are located in the main directory.
Scripts process the data, reproduce figures presented in our article and prepare inputs necessary for the discrete and continous models of HSPC dynamics.

The easiest way to reproduce our scRNA-Seq analysis is to use our [[https://docs.sylabs.io/guides/3.10/user-guide/][Singularity]] container.
The container contains R and python installations and all necessary packages and can be found on [[https://doi.org/10.5281/zenodo.10156542][Zenodo]]

** How to use
To reproduce the analysis presented in the paper:

1. Clone and enter the repository:
#+begin_src bash
git clone https://github.com/Iwo-K/HSPCdynamics
cd HSPCdynamics
#+end_src
2. Download the container from [[https://doi.org/10.5281/zenodo.10156542][Zenodo]]
3. Modify the container variable in the run_analysis.sh script to point to the singularity image
4. Download the input data from [[https://doi.org/10.5281/zenodo.10156542][Zenodo]], decompress it and move it to the directory: HSPCdynamics/data
5. Run the analysis:
 #+begin_src bash
 ./run_analysis.sh
 #+end_src

This script will create several directories to hold output files from respective analysis script, e.g. figures or processed data. Additionally, the .R scripts will be converted into .html reports and .py scripts into Jupyter/IPython notebooks, to allow easy inspection.

* Discrete model
** create_input directory
Prior to running the analysis the necessary input files are generated using the scripts in this directory.
Scripts read in data from the upstream scRNA-Seq analysis.
** scripts in the main directory
Generates the main discrete model of HSPC dynamics, with the following scripts.
For convenience we provide our parameter values in the ~output/best.txt~ file (and with bounds in ~output/bouds.txt~), which are recognised by the matlab scripts.

*** C1main.m
This file finds the best fit for the discrete model, plots and saves the figures, saves the estimated parameters and the model predictions.
Since it takes many iterations and initial guesses to find the best fit, it is currently set to read the best fit already computed.
To start from scratch, delete the best.txt file in the input folder and launch the main many times from differnt starts with your favourite multistart launcher (e.g., PESTO) or just change lines 186 and 174.

*** C2_profile_likhelihood.m
This file computes profiles for each parameter upon blocking the parameter values one at the time and then reoptimizing the remaining parameters.

*** C3_compute_bounds.m
Reads the previously computed profiles to extract lower and upper bounds in the file bounds.txt.

*** C4_switch_off.m
Simulates the system upon switching off any population, passed by user as a vector of cluster label, and saves the figure in the figure folder with corresponding label.

*** C5_waiting_times.m
Calculates time until 1 cell is accumulated in the chosen cluster after initialisation with 1 cell in the cluster(s) 0. In the article we refere to this measure as the journey time.

** biphasic directory
Generates the bi-phasic model of HSPC dynamics. This means we specify one set of parameters (differentiation and net proliferation rates) for the first 4 time-point (the 'recovery' phase) and another set for remaining time-points (the 'homeostasis' phase).

** bootstrap directory
Provides confidence intervals for the discrete model (estimated using bootstrap) and predicted relative cell numbers for the transplantation data (from [[https://www.nature.com/articles/s41556-020-0512-1][Dong et al. 2020]]).

*** Confidence intervals for the discrete models
Scripts C1_bootstrap.m and C2_generate_bounds.m read bootstrap simulations and estimate 95% confidence intervals for predicted relative cell numbers in each cluster over time.

*** Prediction for the transplantation data
Calculated in scripts C3_prepare_input_transplantation_propagate_error.m and C4_tx_estimation_bootstrap.m
** any_topology
The code "main_any_topology" runs an optimisation of our model for an arbitrary topology, independently of the PAGA scheme. The only restriction is that the number of differentiation has to be smaller or equal 302, otherwise the number of unknown variables is bigger than the number of data.

To select the rate that you want to allow, enter 1 in the file "any_topology.xlsx". For example, if you want to allow the rate from cluster 16 to cluster 0c, place 1 in the intersection of the row "cluster 16" and column "cluster 0c". Any undesired rate should be set to 0.

Run the code with multiple starts and check the plot to determine if you find a reasonable minimum (model fits the data well). You can compare models with likelihood ratio or akaike criterion.
** explore_parameter_numbers
This session is meant to explore different topologies, similarly to the any_topology code, but focuses only on the topologies that are contained in our selected PAGA edges.

Since we chose 37 edges, there are in principle 2^37 submodels. Since this is too many combinations, we picked only those combinations that connect all populations with at least one edge and that overall
derive from the HSC tip cluster 0a. They are still many models, so we focussed on models that have less parameters than ours. For some of them we could explore
all the combinations, for some others there was still a computaional burdain. We nevertheles provide the results we obtained,
and a code that would allow to further explore more combinations if desiderd.

Run count_combinations for storing all the combinations that are meaningful (>200000)

Then select the desired number of parameters and print a dataframe with those combinations (check_combinations).

Finally use run_combinations to obtained the desired minima.
** transplantation data model refit - tx_iterative_model_selection
The code launch_loop.m launches an optimisation of the kinetics after transplantation assuming that only a subgroup of populations change their differentiation/proliferation rates. Each run of launch_loop tests 1000 combinations
of populations that may change their rate. If ran 66 times, it covers all the possibilities of choosing a subgroup of changing populations out of the 16 that can possibly change (~66000 combinations).

It stores the best chisquared in the track file for each combination of changed population.

Then, run analyse_track.m to assess the minimum number of parameters that still fits the data (see methods).

Finally run pl_top_models to find confidence bounds on the changing parameters.

* Continuous model
Code uses the pseudodynamics framework devloped by Fabian Theis' lab ([[https://doi.org/10.1038/s41587-019-0088-0][paper]], [[https://github.com/theislab/pseudodynamics][Github]]).
Code requiers a specific version of matlab: r2017b and gcc: gcc-4.9.4-gcc-4.8.5-3sdjf2c.

Analysis is provided in the PD_model directory, each analysed trajectory in a separate subdirectory.
Each directory follows the same scheme with scripts numbered according to the order of execution.
- C1 extracts the trajectory and cell numbers
- C2-5 prepare data and fits the model
- C6-8 plot and write the data

* Note on cluster numbers
In the publication clusters were relabelled with consecutive numbers, this repository uses the original numbers, with the mapping provided below:

#+CAPTION: Cluster name mapping
| python | matlab | original name | publication |
|--------+--------+---------------+-------------|
|      0 |      1 |            0a |          0a |
|      1 |      2 |             1 |           1 |
|      2 |      3 |             2 |           2 |
|      3 |      4 |             3 |           3 |
|      4 |      5 |             4 |           4 |
|      5 |      6 |             5 |           5 |
|      6 |      7 |             6 |           6 |
|      7 |      8 |             7 |           7 |
|      8 |      9 |             8 |           8 |
|      9 |     10 |             9 |           9 |
|     10 |     11 |            10 |          10 |
|     11 |     12 |            11 |          11 |
|     12 |     13 |            12 |          12 |
|     13 |     14 |            14 |          13 |
|     14 |     15 |            16 |          14 |
|     15 |     16 |            20 |          15 |
|     16 |     17 |            24 |          16 |
|     17 |     18 |            25 |          17 |
|     18 |     19 |            26 |          18 |
|     19 |     20 |            28 |          19 |
|     NA |     21 |            0a |          0a |
|     NA |     22 |            0b |          0b |
